// ng-gx-split-text/esm2015/lib/directives/ng-gx-split-text.directive.js
import { Directive, ElementRef, Input } from '@angular/core';
import { defaults } from '../utils/defaults/defaults';
import { defaultOptions } from '../models/default-options';
import { SplitNodes } from '../utils/split-nodes/split-nodes';
export class NgGxSplitTextDirective {
    constructor(el) {
        this.el = el;
        this.options = defaultOptions;
        this.init = false;
    }
    ngOnInit() {
        this.setCurrentOptions();
        this.initSplitNodes();
    }
    ngAfterViewInit() {
        if (!this.currentOptions.defer) {
            this.initSplit();
        }
    }
    initSplit() {
        if (this.init) {
            console.warn('Warning! Text is already initialized');
            return;
        }
        this.init = true;
        this.saveSrcText();
        this.splitNodes.initSplitNodes();
    }
    saveSrcText() {
        this.srcTextContent = this.el.nativeElement.innerHTML;
    }
    get words() {
        return this.splitNodes.words;
    }
    get lineWords() {
        return this.splitNodes.lineWords;
    }
    get chars() {
        if (this.options.onlyWords) {
            console.warn('Warning! You chose only words.');
            return;
        }
        return this.splitNodes.chars;
    }
    get lineChars() {
        if (this.options.onlyWords) {
            console.warn('Warning! You chose only words.');
            return;
        }
        return this.splitNodes.lineChars;
    }
    get nativeElement() {
        return this.el.nativeElement;
    }
    get srcText() {
        return this.srcTextContent;
    }
    get isInit() {
        return this.init;
    }
    resetSplit() {
        this.nativeElement.innerHTML = this.srcText;
    }
    setCurrentOptions() {
        this.currentOptions = defaults(this.options, defaultOptions);
    }
    initSplitNodes() {
        this.splitNodes = new SplitNodes(this.el.nativeElement.textContent, this.el.nativeElement, this.currentOptions);
    }
}
NgGxSplitTextDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngGxSplitText], ngGxSplitText'
            },] }
];
NgGxSplitTextDirective.ctorParameters = () => [
    { type: ElementRef }
];
NgGxSplitTextDirective.propDecorators = {
    options: [{ type: Input, args: ['splitOptions',] }]
};
